url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/game/SOCBoard.java

```diff
@@ -0,0 +1,2512 @@
+        final int[] numPath_6pl =
+        {
+           15, 9, 4, 0, 1, 2, 7, 13, 20, 26, 31, 35, 34, 33, 28, 22,  // outermost hexes
+           16, 10, 5, 6, 12, 19, 25, 30, 29, 23,  // middle ring of hexes
+           17, 11, 18, 24                         // center hexes
+        };
+
+        SOCGameOption opt_breakClumps = (opts != null ? (SOCGameOption)opts.get("BC") : null);
+
+        // shuffle and place the land hexes, numbers, and robber:
+        // sets robberHex, contents of hexLayout[] and numberLayout[].
+        // Also checks vs game option BC: Break up clumps of # or more same-type hexes/ports
+        {
+            final int[] landHex = is6player ? landHex_6pl : landHex_v1;
+            final int[] numPath = is6player ? numPath_6pl : numPath_v1;
+            final int[] numbers = is6player ? number_6pl : number_v1;
+            makeNewBoard_placeHexes
+                (landHex, numPath, numbers, opt_breakClumps);
+        }
+
+        // copy and shuffle the ports, and check vs game option BC
+        final int[] portTypes = (is6player) ? PORTS_TYPE_V2 : PORTS_TYPE_V1;
+        int[] portHex = new int[portTypes.length];
+    	System.arraycopy(portTypes, 0, portHex, 0, portTypes.length);
+        makeNewBoard_shufflePorts
+            (portHex, opt_breakClumps);
+        if (is6player)
+            portsLayout = portHex;  // No need to remember for 4-player standard layout
+
+        /*
+        // place the ports (hex numbers and facing) within hexLayout and nodeIDtoPortType
+
+        placePort(portHex[0], 0, 3);  // Facing 3 is SE: see hexLayout's javadoc.
+                                      //   0 is hex number (index within hexLayout)
+        placePort(portHex[1], 2, 4);  // Facing 4 is SW, at hex number 2
+        placePort(portHex[2], 8, 4);  // SW
+        placePort(portHex[3], 21, 5); // W
+        placePort(portHex[4], 32, 6); // NW
+        placePort(portHex[5], 35, 6); // NW
+        placePort(portHex[6], 33, 1); // NE
+        placePort(portHex[7], 22, 2); // E
+        placePort(portHex[8], 9, 2);  // E
+        */
+
+        // place the ports (hex numbers and facing) within hexLayout and nodeIDtoPortType.
+        // fill out the ports[] vectors with node coordinates where a trade port can be placed.
+        nodeIDtoPortType = new int[MAXNODEPLUSONE];
+        for (int i = 0; i <= MAXNODE; ++i)
+            nodeIDtoPortType[i] = -1;  // -1 means not a port (or not a valid node coord)
+        if (is6player)
+        {
+            for (int i = 0; i < PORTS_FACING_V2.length; ++i)
+            {
+                final int ptype = portHex[i];
+                final int[] nodes = getAdjacentNodesToEdge_arr(PORTS_EDGE_V2[i]);
+                placePort(ptype, -1, PORTS_FACING_V2[i], nodes[0], nodes[1]);
+                ports[ptype].addElement(new Integer(nodes[0]));
+                ports[ptype].addElement(new Integer(nodes[1]));
+            }            
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 843:850, 1389:1396.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.